types/src/waypoint.rs

```diff
@@ -144,6 +144,45 @@ impl Ledger2WaypointConverter {
     }
 }
 
+impl<'de> Deserialize<'de> for Waypoint {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            let s = <String>::deserialize(deserializer)?;
+            Waypoint::from_str(&s).map_err(D::Error::custom)
+        } else {
+            // In order to preserve the Serde data model and help analysis tools,
+            // make sure to wrap our value in a container with the same name
+            // as the original type.
+            #[derive(::serde::Deserialize)]
+            #[serde(rename = "Waypoint")]
+            struct Value(Version, HashValue);
+
+            let value = Value::deserialize(deserializer)?;
+            Ok(Waypoint {
+                version: value.0,
+                value: value.1,
+            })
+        }
+    }
+}
+
+impl Serialize for Waypoint {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_human_readable() {
+            self.to_string().serialize(serializer)
+        } else {
+            // See comment in deserialize.
+            serializer.serialize_newtype_struct("Waypoint", &(self.version, self.value))
```

Comment:
Isn't this a tuple struct?

Reply:
This might be an LCS specific hack, but I saw other cases, but effectively a tuple is an anonymous struct.